receiver/postgresqlreceiver/client.go

```diff
@@ -0,0 +1,285 @@
+}
+
+func (c *postgreSQLClient) getDatabaseSize(ctx context.Context, databases []string) ([]MetricStat, error) {
+	query := filterQueryByDatabases("SELECT datname, pg_database_size(datname) FROM pg_catalog.pg_database WHERE datistemplate = false", databases, false)
+
+	return c.collectStatsFromQuery(ctx, query, true, false, "db_size")
+}
+
+func (c *postgreSQLClient) getDatabaseTableMetrics(ctx context.Context) ([]MetricStat, error) {
+	query := `SELECT schemaname || '.' || relname AS table,
+	n_live_tup AS live,
+	n_dead_tup AS dead,
+	n_tup_ins AS ins,
+	n_tup_upd AS upd,
+	n_tup_del AS del,
+	n_tup_hot_upd AS hot_upd
+	FROM pg_stat_user_tables;`
+
+	return c.collectStatsFromQuery(ctx, query, false, true, "live", "dead", "ins", "upd", "del", "hot_upd")
+}
+
+func (c *postgreSQLClient) getBlocksReadByTable(ctx context.Context) ([]MetricStat, error) {
+	query := `SELECT schemaname || '.' || relname AS table, 
+	coalesce(heap_blks_read, 0) AS heap_read, 
+	coalesce(heap_blks_hit, 0) AS heap_hit, 
+	coalesce(idx_blks_read, 0) AS idx_read, 
+	coalesce(idx_blks_hit, 0) AS idx_hit, 
+	coalesce(toast_blks_read, 0) AS toast_read, 
+	coalesce(toast_blks_hit, 0) AS toast_hit, 
+	coalesce(tidx_blks_read, 0) AS tidx_read, 
+	coalesce(tidx_blks_hit, 0) AS tidx_hit 
+	FROM pg_statio_user_tables;`
+
+	return c.collectStatsFromQuery(ctx, query, false, true, "heap_read", "heap_hit", "idx_read", "idx_hit", "toast_read", "toast_hit", "tidx_read", "tidx_hit")
+}
+
+func (c *postgreSQLClient) collectStatsFromQuery(ctx context.Context, query string, includeDatabase bool, includeTable bool, orderedFields ...string) ([]MetricStat, error) {
+	rows, err := c.client.QueryContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	errors := scrapererror.ScrapeErrors{}
+	metricStats := []MetricStat{}
+	for rows.Next() {
+		rowFields := make([]interface{}, 0)
+
+		// Build a list of addresses that rows.Scan will load column data into
+		if includeDatabase {
+			var val string
+			rowFields = append(rowFields, &val)
+		}
+		if includeTable {
+			var val string
+			rowFields = append(rowFields, &val)
+		}
+		for range orderedFields {
+			var val string
+			rowFields = append(rowFields, &val)
+		}
+
+		if err := rows.Scan(rowFields...); err != nil {
+			return nil, err
+		}
+
+		database := c.database
+		if includeDatabase {
+			if v, err := convertInterfaceToString(rowFields[0]); err != nil {
+				errors.AddPartial(len(rowFields), err)
+				continue
+			} else {
+				database = v
+			}
+			rowFields = rowFields[1:]
+		}
+		table := ""
+		if includeTable {
+			if v, err := convertInterfaceToString(rowFields[0]); err != nil {
+				errors.AddPartial(len(rowFields), err)
+				continue
+			} else {
+				table = v
+			}
+			rowFields = rowFields[1:]
+		}
+
+		stats := map[string]string{}
+		for idx, val := range rowFields {
+			if v, err := convertInterfaceToString(val); err != nil {
+				errors.AddPartial(1, err)
```

Comment:
Since we don't assign a value here, are we relying on the numeric parsing operation to fail, thereby omitting the data point?

Reply:
The data point is omitted either due to the data not being added to stats (when we iterate like `for k, v := range table.stats {`) or will be skipped by checking that the field exists before trying to parse (when we're iterating over a known list of stats). First one already works, second will require a small change.

Reply:
👍 No need to implement the second. I just missed how it was catching it. Thanks for clarifying.